node/src/components/block_synchronizer/block_builder.rs

```diff
@@ -233,7 +263,47 @@ impl BlockBuilder {
             error!(%error, "register block execution enqueued failed");
             self.abort()
         } else {
-            self.touch();
+            match (
+                self.should_fetch_execution_state,
+                self.execution_progress.start(),
+            ) {
+                (true, _) => {
+                    let block_hash = self.block_hash();
+                    error!(%block_hash, "invalid attempt to start block execution on historical block");
+                    self.abort();
+                }
+                (false, None) => {
+                    let block_hash = self.block_hash();
+                    error!(%block_hash, "invalid attempt to start block execution");
+                    self.abort();
+                }
+                (false, Some(executing_progress)) => {
+                    self.touch();
+                    self.execution_progress = executing_progress;
+                }
+            }
```

Comment:
**Nit**: I think the semantics of `start()` are a bit hard to read. Since we can only start from `ExecutionProgress::Idle`, I think we can remove `start()` and replace with something like:
```suggestion
            match (
                self.should_fetch_execution_state,
                self.execution_progress,
            ) {
                (true, _) => {
                    let block_hash = self.block_hash();
                    error!(%block_hash, "invalid attempt to start block execution on historical block");
                    self.abort();
                }
                (false, ExecutionProgress::Idle) => {
                    self.touch();
                    self.execution_progress = ExecutionProgress::Started;
                }
                (false, _) => {
                    let block_hash = self.block_hash();
                    error!(%block_hash, execution_progress=%self.execution_progress "cannot track block execution progress");
                    self.abort();
                }
            }
```